sm_ipt/include/sm_ipt.h

```diff
@@ -0,0 +1,141 @@
+ *
+ * @param packet Received data.
+ * @param len Length of received data.
+ */
+typedef void (*sm_ipt_receive_handler_t)(const uint8_t *packet, size_t len);
+
+/** @brief Single shared memory queue.
+ *
+ * Used by @ref sm_ipt_ctx to store queue information.
+ *
+ * Fields of this structre are used internally and should not
+ * be used by the user.
+ */
+struct sm_ipt_queue {
+	/** @brief Maximum allocable shared memory size. */
+	size_t allocable_size;
+
+	/** @brief Size of a single allocable block. */
+	size_t block_size;
+
+	/** @brief Queue placed in shared memory. */
+	struct sm_ipt_shm_queue *shm_queue;
+};
+
+/** @brief SM IPT context.
+ *
+ * Used to store current context data.
+ *
+ * Fields of this structre are used internally and should not
+ * be used by the user.
+ */
+struct sm_ipt_ctx {
+	/** @brief Input queue. */
+	struct sm_ipt_queue in;
+
+	/** @brief Output queue. */
+	struct sm_ipt_queue out;
+
+	/** @brief Output semaphore for allocation. */
+	sm_ipt_os_sem_t out_sem;
+
+	/** @brief Output mutex for sending. */
+	sm_ipt_os_mutex_t out_mutex;
+
+	/** @brief Allocation bitmap. */
+	sm_ipt_os_atomic_t free_mask[IS_ENABLED(CONFIG_SM_IPT_NUM_BLOCKS_64) ? 2 : 1];
+
+	/** @brief Receive handler. */
+	sm_ipt_receive_handler_t receive_handler;
+
+	/** @brief OS Context - to be implemented by OS layer. */
+	struct sm_ipt_os_ctx os_ctx;
+};
+
+/** @brief Initialize the SM IPT
+ *
+ * This function initializes shared memory and context data.
+ *
+ * @param ctx         SM IPT Context.
+ * @param callback    Callback to receive handler.
+ *
+ * @return            0 on success or negative error code.
+ */
+int sm_ipt_init(struct sm_ipt_ctx *ctx, sm_ipt_receive_handler_t callback);
+
+/** @brief Free SM IPT rx buffer
+ *
+ * This function frees shared memory buffer for receiving data.
+ *
+ * @param ctx         SM IPT Context.
+ * @param packet      Selected buffer.
+ */
+void sm_ipt_free_rx_buf(struct sm_ipt_ctx *ctx, const uint8_t *buf);
+
+/** @brief Allocate SM IPT tx buffer
+ *
+ * This function allocates shared memory buffer for trasmitting data.
+ *
+ * @param ctx         SM IPT Context.
+ * @param packet      Pointer to selected buffer.
+ * @param len         Length of allocation.
+ */
+void sm_ipt_alloc_tx_buf(struct sm_ipt_ctx *ctx, uint8_t **buf, size_t len);
+
+/** @brief Free SM IPT tx buffer
+ *
+ * This function frees shared memory buffer for trasmitting data.
+ *
+ * @param ctx         SM IPT Context.
+ * @param packet      Pointer to selected buffer.
+ */
+void sm_ipt_free_tx_buf(struct sm_ipt_ctx *ctx, uint8_t *buf);
+
+/** @brief Send message through SM IPT
+ *
+ * This function sends data stored at buf in shared memory.
+ *
+ * @param ctx         SM IPT Context.
+ * @param packet      Selected buffer.
+ * @param len         Length of message.
+ *
+ * @return            0 on success or negative error code.
+ */
+int sm_ipt_send(struct sm_ipt_ctx *ctx, uint8_t *buf, size_t len);
```

Comment:
Can buf be const ?
```suggestion
int sm_ipt_send(struct sm_ipt_ctx *ctx, const uint8_t *buf, size_t len);
```

Reply:
The function changes the content just before sending (the size of the allocated block), so it should be without `const`.